### PR TITLE
Added energy, fuel, and heat attributes for resistances

### PR DESCRIPTION
**Pull request by Amazinite**
Ref: #4517

I *think* this works. It compiles, at least. I implemented this by using a modified DoRepair function that I called DoStatusEffect. It first calculates how much resistance could possibly be needed given what the resistance attribute and current status effect are, then limits the needed resistance further by energy and fuel costs if necessary. Then, given how much resistance was done, it updates the energy, fuel, and heat of the ship. **Resistance attributes will not be applied if the ship is disabled,** just like shield generation and hull repair. This is the opposite of how resistance is currently applied, where it will affect the ship if disabled.

Given that I did this using a modified DoRepair function, and DoRepair is used to repair carried ships after the mothership is repaired, I left a TODO comment asking if the mothership should also impart its left over resistance attributes onto the carried ships. I'm pretty sure the DoResistance function would need to be altered slightly to allow that, though.

### CODE REVIEW

**Comment by Amazinite** [comment]
Note that these attributes are not displayed anywhere on the ship stats because there's no room for them. Ref: #4215.

File: source/Ship.cpp [hunk 1]
```diff
@@ -86,6 +86,39 @@ namespace {
...
+		if(fuelCost)
+			resistance = min(resistance, fuel / fuelCost);
+		
+		// Update the energy, fuel, and heat given how much resistance is being used,
+		// then update the stat.
+		if(resistance > 0.)
+		{
+			energy -= resistance * energyCost;
+			fuel -= resistance * fuelCost;
+			heat += resistance * heatCost;
```
Comment (tehhowch) [inline source/Ship.cpp]: We should account for negative heatCost, i.e. being required to supply heat to use the resistance.

File: source/Ship.cpp [hunk 2]
```diff
@@ -1625,13 +1658,35 @@ void Ship::DoGeneration()
 		if(shieldsAvailable)
 			heat += (shieldsAvailable - shieldsRemaining) * shieldsHeat;
 	}
+	
 	// Handle ionization effects, etc.
+	// TODO: Mothership gives status resistance to carried ships?
 	if(ionization)
-		ionization = max(0., .99 * ionization - attributes.Get("ion resistance"));
+	{
+		const double ionResistance = attributes.Get("ion resistance");
```
Comment (tehhowch) [inline source/Ship.cpp]: Making these const here, but not in the method called with them, is meaningless. With the new reduced scope, you can discard the const qualifier (or update the called method's signature).

**Comment by Zitchas** [comment]
If the other resistances apply even when a ship is disabled, why shouldn't these ones?

Ideally, I'd say there should be a companion attribute that determines whether the resistance does or does not apply. I think probably the ideal would be to have this companion value be a multiplier that is applied to the resistance value when disabled. The default value would be 0. So, when disabled, the resistance stops applying. Any passive thing that doesn't need a functioning ship to keep working would have that value set as 1. If a thing is some part passive, part active thing that loses some functionality but stays functional, it could have a value somewhere in between. Negative values could indicate a device that malfunctions (or a ship with an innate weakness that becomes vulnerable when disabled) could have a negative value. A ship that for whatever reason is less vulnerable to something when disabled could have a value greater than 1.

From a lore perspective, if something is an active system, (say, some kind of field the ship projects to reduce the effect of some kind of damage), then yes, it shouldn't apply when disabled. But if it is a passive system (or specifically designed to keep functioning without the support of the rest of ship), then it should keep functioning.

It seems rather arbitrary to decide in the mechanics that any particular resistance must or must not be an active or passive system, when realistically they could be either.

**Comment by Amazinite** [comment]
> If the other resistances apply even when a ship is disabled, why shouldn't these ones?

The only other resistance-like attributes that exist outside of these here are are shield generation, hull repair, and cooling, and the former two of those don't work if disabled, so I'm not sure what you mean by "the other resistances."

**Comment by Zitchas** [comment]
I think I misread something in your earlier post about some resistances working when disabled, but these ones won't work when disabled.

**Comment by Amazinite** [comment]
It's the case that right now in the master, these three resistances are passive (i.e. work when disabled), but this PR makes them active (i.e. must be enabled).

Give me a few minutes though. A few extra lines and I can have attributes determine that. (Although, I believe you'll get bad results if you mix the costs with passive resistances, as I attempted to create passive hull repair/shield generation once and it didn't work.)

**Comment by Amazinite** [comment]
Damn is this cutting it close.
image
Any wording that would make this shorter?

**Comment by Zitchas** [comment]
Resist disruption while disabled is slightly shorter.

**Comment by Amazinite** [comment]
I want to avoid that wording as it suggests that the effect is only resisted when disabled. "Even if" makes it clear that the resistance is applied while enabled, but while disabled as well.

**Comment by Zitchas** [comment]
Yes, that does make sense. "Even if" is pretty clear. (Even if is also shorter, I think. Your picture shows "even when" though)

**Comment by Amazinite** [comment]
For whatever reason, GitHub doesn't let you link issues and PRs sometimes. The search bar just doesn't show the issue or PR you want to select. Randomly commenting here to see if it's based on activity and only shows some number of recently active issues and PRs.

**Comment by Amazinite** [comment]
That answers that question.

**Pull request merged**